src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensSingleRootFileViewProvider.java

```diff
@@ -0,0 +1,184 @@
+                                                          @NotNull VirtualFile virtualFile,
+                                                          boolean eventSystemEnabled) {
+        super(manager, virtualFile, eventSystemEnabled);
+        this.helper = new LSPSemanticTokensFileViewProviderHelper(this);
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return helper.isEnabled();
+    }
+
+    @Override
+    public boolean isKeyword(int offset) {
+        return helper.isKeyword(offset);
+    }
+
+    @Override
+    public boolean isOperator(int offset) {
+        return helper.isOperator(offset);
+    }
+
+    @Override
+    public boolean isStringLiteral(int offset) {
+        return helper.isStringLiteral(offset);
+    }
+
+    @Override
+    public boolean isNumericLiteral(int offset) {
+        return helper.isNumericLiteral(offset);
+    }
+
+    @Override
+    public boolean isRegularExpression(int offset) {
+        return helper.isRegularExpression(offset);
+    }
+
+    @Override
+    public boolean isComment(int offset) {
+        return helper.isComment(offset);
+    }
+
+    @Override
+    public boolean isDeclaration(int offset) {
+        return helper.isDeclaration(offset);
+    }
+
+    @Override
+    public boolean isReference(int offset) {
+        return helper.isReference(offset);
+    }
+
+    @Override
+    @NotNull
+    public ThreeState isType(int offset) {
+        return helper.isType(offset);
+    }
+
+    @Override
+    @Nullable
+    public String getElementDescription(@NotNull PsiElement element, @Nullable PsiElement referenceElement) {
+        return helper.getElementDescription(element, referenceElement);
+    }
+
+    @Override
+    @Nullable
+    public TextRange getSemanticTokenTextRange(int offset) {
+        return helper.getSemanticTokenTextRange(offset);
+    }
+
+    @Override
+    public void addSemanticToken(@NotNull TextRange textRange,
+                                 @Nullable String tokenType,
+                                 @Nullable List<String> tokenModifiers) {
+        helper.addSemanticToken(textRange, tokenType, tokenModifiers);
+    }
+
+    /**
+     * Returns the semantic token for the offset.
+     *
+     * @param offset the offset
+     * @return the semantic token or null if no semantic token exists at the offset
+     */
+    @Nullable
+    LSPSemanticToken getSemanticToken(int offset) {
+        return helper.getSemanticToken(offset);
+    }
+
+    @Nullable
+    protected PsiReference getSemanticTokenReference(int offset) {
+        LSPSemanticToken semanticToken = isEnabled() ? getSemanticToken(offset) : null;
+        LSPSemanticTokenElementType elementType = semanticToken != null ? semanticToken.getElementType() : null;
+        return elementType == LSPSemanticTokenElementType.REFERENCE ? new LSPSemanticTokenPsiReference(semanticToken) : null;
+    }
+
+    @Override
+    public PsiReference findReferenceAt(int offset) {
```

Comment:
These help ensure that first-class references for the file/element/language are respected while adding semantic tokens-based references if present.